Fix #70767

xldenis opened the pull request:
This PR changes the format of MIR dump filenames to include the crate name rather than `rustc` at the start. 

As a result, we can now place mir-opt tests in the same directory as the source files, like with UI tests. I had to make sure that `compiletest` added a bit_width suffix to the expected files when appropriate but otherwise the change is only moving the files to the correct location and ensuring that the `EMIT_MIR` lines are correct.

Fixes #70767
cc @oli-obk

xldenis committed:
add crate name to mir dumps

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

xldenis commented:
ah seems like i didn't generate the 32bit wide expectations. I'll have to run bless with a 32bit target i guess

oli-obk commented:
r? @oli-obk

xldenis committed:
Move mir-opt tests to toplevel

xldenis commented:
I wasn't able to run 32 bit tests (mac) but I moved the test expectations manually to the top level so it should be good. We'll see if CI agrees.

oli-obk commented on src/test/mir-opt/const_prop/control_flow_simplification.hello.ConstProp.diff.all:
```diff
@@ -41,19 +41,19 @@
                                            // + val: Value(Scalar(<ZST>))
                                            // mir::Constant
-                                           // + span: $SRC_DIR/std/src/macros.rs:LL:COL
+                                           // + span: $LIB_DIRrary/std/src/macros.rs:13:23: 13:46
```
This regressed  (probably during the rebase over the latest directory restructuring)

xldenis replied:
wow... what a bizarre diff, i'll go fix it

oli-obk commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3124,15 +3124,17 @@ impl<'test> TestCx<'test> {
 
         if self.config.bless {
-            let _ = std::fs::remove_dir_all(&test_dir);
+            let _ = std::fs::remove_dir_all(&self.testpaths.file.with_extension(""));
```
huh, this does anything? don't we need to glob all such files?

xldenis replied:
that needs to be removed. it was to clean up the old files during the transition.

oli-obk replied:
The problem is that without any removal, we'll end up with orphaned files (e.g. when the `// EMIT_MIR` command is changed, the old files would stay). But with the new scheme, we can't just remove all `filename.*` where `*` is not `rs`, because 32 bit and 64 bit generation happens in different runs, so we'd keep removing the files from the other bit width.

I'm not sure what to do here, except maybe ignore all files that don't have the bitwidth that is currently being tested.

xldenis replied:
it's already the case that we had orphaned files so the previous code wasn't bullet proof. With the current code can't we just remove `expected_file`?

xldenis replied:
ah no I see, you mean if we change the name of the tested function we won't clean up the old expectations. hmmm.

xldenis replied:
we could nuke everything with the same crate name? it's an over approximation but not too bad?

xldenis replied:
same problem as removing the crate name in `EMIT_MIR` how do we know which files belong to which tests? Especially when they don't necessarily start with that test's crate prefix, ie: `src/test/mir-opt/core.ops-function-Fn-call.AddMovesForPackedDrops.before.mir.all`

oli-obk replied:
O_o I did not think of that. I guess we need to prefix the name to everything that doesn't already have it prefixed?

xldenis replied:
yea i guess so

xldenis replied:
though there is the problem of knowing _what_ to prefix :/ again, the files start with crate names which are different from the source file names.

xldenis replied:
ok I've used the `glob` crate to remove all files starting with the crate prefix of the test. The next step would be to prefix expectations that are from other crates.

xldenis replied:
done! i made sure that all files are prefixed with the cratename.

oli-obk commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3124,15 +3124,17 @@ impl<'test> TestCx<'test> {
-        let mut test_dir = self.testpaths.file.with_extension("");
+        let test_dir = self.testpaths.file.parent().unwrap();
 
+        let mut bit_width = "all";
```
keep this string empty and prepend a `.` to the string if there's an actual bit width. This way we have shorter names in the regular case

xldenis replied:
👍

oli-obk commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3216,8 +3199,9 @@ impl<'test> TestCx<'test> {
+
                 if self.config.bless {
-                    let _ = std::fs::create_dir_all(&test_dir);
+                    // let _ = std::fs::create_dir_all(&test_dir);
```
leftover commented out code

oli-obk commented on src/test/mir-opt/while-storage.rs:
```diff
@@ -5,7 +5,7 @@ fn get_bool(c: bool) -> bool {
 }
 
-// EMIT_MIR rustc.while_loop.PreCodegen.after.mir
+// EMIT_MIR while_storage.while_loop.PreCodegen.after.mir
```
do you think it's doable with reasonable effort to not have to specify the prefix and instead have the test suite prepend the file name (without `.rs`) automatically?

xldenis replied:
if i can find the function in rustc which turns a filename into a cratename, it should be possible (note the file is `while-storage` but the crate is `while_storage`)

xldenis replied:
oh actually, I don't think it's super easily possible, because sometimes the expectations are from `core` and other crates... which we can't predict easily.

oli-obk replied:
jup makes sense, let's just go with the status quo

oli-obk commented on src/test/mir-opt/array_index_is_temporary.main.SimplifyCfg-elaborate-drops.after.mir.32bit:
```diff
@@ -0,0 +1,97 @@
+// MIR for `main` after SimplifyCfg-elaborate-drops
```
This file is almost a duplicate of another file, the only difference is the `_` vs `-` in the main name

xldenis replied:
i moved these by hand so there's probably a couple errors hiding still.

xldenis commented:
@oli-obk tests are finally green ✅ and I think I've addressed all your comments. I think this should be merged quickly before another test is merged and breaks it all over again

oli-obk commented:
@bors r+ p=6 bitrotty

bors commented:
:pushpin: Commit ab5bf2bfcc53caf0fb34cf6c622c4dff7ac8a8ce has been approved by `oli-obk`

<!-- @bors r=oli-obk ab5bf2bfcc53caf0fb34cf6c622c4dff7ac8a8ce -->
<!-- homu: {"type":"Approved","sha":"ab5bf2bfcc53caf0fb34cf6c622c4dff7ac8a8ce","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit ab5bf2bfcc53caf0fb34cf6c622c4dff7ac8a8ce with merge 911d21f61a2559c6649ecead6def55ef2206bbbf...
<!-- homu: {"type":"BuildStarted","head_sha":"ab5bf2bfcc53caf0fb34cf6c622c4dff7ac8a8ce","merge_sha":"911d21f61a2559c6649ecead6def55ef2206bbbf"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/922942516)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/922942516","builder_name":"checks-actions"} -->

xldenis commented:
ah a commit was already merged which broke this PR :/

xldenis commented:
I added the new `InstrumentCoverage` test @oli-obk that was the only failure on merge so this should pass 🤞

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 7b2bf1fe68e4a4c9e3ea4d790cfa6e57c8f1ce3d has been approved by `oli-obk`

<!-- @bors r=oli-obk 7b2bf1fe68e4a4c9e3ea4d790cfa6e57c8f1ce3d -->
<!-- homu: {"type":"Approved","sha":"7b2bf1fe68e4a4c9e3ea4d790cfa6e57c8f1ce3d","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 7b2bf1fe68e4a4c9e3ea4d790cfa6e57c8f1ce3d with merge ed0bd1a3dacfb34205d3bc21052a04c555e9d283...
<!-- homu: {"type":"BuildStarted","head_sha":"7b2bf1fe68e4a4c9e3ea4d790cfa6e57c8f1ce3d","merge_sha":"ed0bd1a3dacfb34205d3bc21052a04c555e9d283"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/923408523)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/923408523","builder_name":"checks-actions"} -->

xldenis commented:
uh... there's another failure on that test, seemingly the expected result has changed???

```
failures:

---- [mir-opt] mir-opt/instrument_coverage.rs stdout ----
11        bb0: {
12  -         falseUnwind -> [real: bb1, cleanup: bb2]; // scope 0 at $DIR/instrument_coverage.rs:10:5: 14:6
13  +         StorageLive(_4);                 // scope 0 at $DIR/instrument_coverage.rs:9:11: 9:11
- +         _4 = const std::intrinsics::count_code_region(const 16004455475339839479_u64, const 0_u32, const 397_u32, const 465_u32) -> bb7; // scope 0 at $DIR/instrument_coverage.rs:9:11: 9:11
+ +         _4 = const std::intrinsics::count_code_region(const 16004455475339839479_u64, const 0_u32, const 425_u32, const 493_u32) -> bb7; // scope 0 at $DIR/instrument_coverage.rs:9:11: 9:11
```

I don't understand how it could have passed in branch CI but not on master?

oli-obk commented:
Maybe due to the changed byte offsets in the file due to your comment changes? Did you rebase and re-bless?

xldenis commented:
Yea, locally it seems to ignore the test when I bless because I don’t have a profiler set up, I’m still confused why branch CI passed if the test is wrong.

oli-obk commented:
Oh it's that test -.-. Yea, PR CI doesn't run a profiler either, only bors does that, and only on a single builder. Are you able to set up a profiler or should I try?

xldenis commented:
I'm building with profiler support right now :)

xldenis commented:
Rebuilt and blessed with profiling!

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit f07607f47ab40837c92035e35602731cc80baa81 has been approved by `oli-obk`

<!-- @bors r=oli-obk f07607f47ab40837c92035e35602731cc80baa81 -->
<!-- homu: {"type":"Approved","sha":"f07607f47ab40837c92035e35602731cc80baa81","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit f07607f47ab40837c92035e35602731cc80baa81 with merge af37ac1f38010f0dd2439d552bf56ec22c709bae...
<!-- homu: {"type":"BuildStarted","head_sha":"f07607f47ab40837c92035e35602731cc80baa81","merge_sha":"af37ac1f38010f0dd2439d552bf56ec22c709bae"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35676)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35676","builder_name":"checks-azure"} -->

oli-obk commented:
@bors retry network timeout

bors commented:
:hourglass: Testing commit f07607f47ab40837c92035e35602731cc80baa81 with merge 8611e526b766ce188dc29bd49bf66ff17764ceea...
<!-- homu: {"type":"BuildStarted","head_sha":"f07607f47ab40837c92035e35602731cc80baa81","merge_sha":"8611e526b766ce188dc29bd49bf66ff17764ceea"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/924626417), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35678)
Approved by: oli-obk
Pushing 8611e526b766ce188dc29bd49bf66ff17764ceea to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35678","checks-actions":"https://github.com/rust-lang-ci/rust/runs/924626417"},"merge_sha":"8611e526b766ce188dc29bd49bf66ff17764ceea"} -->
